fix(android): Refactored activity/fragment restore handling

jquick-axway opened the pull request:
**JIRA:**
- https://jira.appcelerator.org/browse/TIMOB-26966
- https://jira.appcelerator.org/browse/TIMOB-17089
- https://jira.appcelerator.org/browse/TIMOB-26890
- https://jira.appcelerator.org/browse/TIMOB-26964
- https://jira.appcelerator.org/browse/TIMOB-26914
- https://jira.appcelerator.org/browse/TIMOB-26975
- https://jira.appcelerator.org/browse/TIMOB-26978
- https://jira.appcelerator.org/browse/TIMOB-15829
- https://jira.appcelerator.org/browse/TIMOB-26989
- https://jira.appcelerator.org/browse/TIMOB-26996

**Summary:**
- [TIMOB-26966](https://jira.appcelerator.org/browse/TIMOB-26966) Fixed bug where back navigating from child window causes app to exit when "Don't keep activities" is enabled as of 7.2.1.
- [TIMOB-26914](https://jira.appcelerator.org/browse/TIMOB-26914) Fixed 8.0.0 regression where an OS forced quit app displays a blank window upon relaunch. (Was incorrectly restoring child activity.)
- [TIMOB-17089](https://jira.appcelerator.org/browse/TIMOB-17089) Resolved view ID conflicts causes app crashes when activity/fragment is restored.
- [TIMOB-26964](https://jira.appcelerator.org/browse/TIMOB-26964) Fixed bug where TabGroup would be blank (no tabs) when calling child window's close() method while "Don't keep activities" is enabled.
- [TIMOB-26890](https://jira.appcelerator.org/browse/TIMOB-26890) Fixed bug where tapping 4th tab in TabGroup can cause a crash due to view ID conflict during fragment restore.
- [TIMOB-26975](https://jira.appcelerator.org/browse/TIMOB-26975) Fixed bug where LiveView does not restart app if "Don't keep activities" is enabled as of 8.0.0.
- [TIMOB-26978](https://jira.appcelerator.org/browse/TIMOB-26978) Added several safety mechanisms to `AlertDialog` and `ProgressIndicator` dialog code to avoid crashes when activity is destroyed, but not finished. (Can happen when "Don't keep activities" is enabled.)
- [TIMOB-15829](https://jira.appcelerator.org/browse/TIMOB-15829) Fixed crash caused by progress indicator set up on the status bar if shown after hiding twice.
- [TIMOB-26989](https://jira.appcelerator.org/browse/TIMOB-26989) Fixed crash caused by restoring an activity having a `TableView` with a `SearchBar` or `SearchView` attached to it. (Proxies should not be released unless activity is finished.)
- [TIMOB-26996](https://jira.appcelerator.org/browse/TIMOB-26996) Fixed bug where Window/TabGroup `close()` method call was ignored until "open" event was fired.
- Increased TabGroup's ViewPager offscreen page limit from 1 to 128 to avoid tab fragments from being destroyed/restored.

---
**Test Restore After OS Force Quit:**
1. On Android, go to its "Developer Options" screen.
2. Tap on the "Background process limit" row. (It's near the bottom of the list.)
3. Tap on "No background processes". _(This tells the OS to force quit the last displayed app when launching another app.)_
4. Build and run the below code on the same Android device.
5. In the app, tap on the "Show Child Window" button.
6. Press the Android "Home" button.
7. Tap on another app such as "Gmail".
8. Press the Android "Home" button.
9. Resume the Titanium app.
10. Verify that the app restarts, starting with the splash screen. _(It will launch with a blank window, but quickly restart. Unfortunately, we are stuck with this behavior.)_
11. Change device's "Background process limit" setting back to "Standard".

```javascript
var window = Ti.UI.createWindow();
var openButton = Ti.UI.createButton({ title: "Show Child Window" });
openButton.addEventListener("click", function(e) {
	var childWindow = Ti.UI.createWindow({ backgroundColor: "orange" });
	var closeButton = Ti.UI.createButton({ title: "Close Me" });
	closeButton.addEventListener("click", function(e) {
		childWindow.close();
	});
	childWindow.add(closeButton);
	childWindow.open();
});
window.add(openButton);
window.open();
```

---
**Test Intent ACTION_VIEW with "Don't keep activities":**
1. Go to the Android device's main "Settings" screen.
2. Tap on "System" under "Settings.
3. Tap on "Developer options" under "System" settings.
4. Enable "Don't keep activities", which should be near the bottom of the list.
5. Build and run the below code on the above Android device.
6. Tap the "Send Intent" button.
7. Verify that a screenshot of the "Send Intent" button is shown in a new window.
8. Tap the Android "Home" button.
9. Resume the app.
10. Verify that the screenshot is still shown.
11. Tap the Android "Back" button.
12. Verify that the it navigated back to the Titanium app with the "Send Intent" button. _(Should no longer exit the app, which was the bug.)_
13. Tap the Android "Back" button.
14. Verify that it exits the app. You should not see a splash screen.

```javascript
var window = Ti.UI.createWindow();
var button = Ti.UI.createButton({
	title: "Send Intent",
});
button.addEventListener("click", function(e){
	button.toImage(function(imageBlob) {
		if (!imageBlob) {
			return;
		}
		var targetFile = Ti.Filesystem.getFile(Ti.Filesystem.applicationDataDirectory, "image.png");
		targetFile.write(imageBlob, false);
		var intent = Ti.Android.createIntent({
			action: Ti.Android.ACTION_VIEW,
			data: targetFile.nativePath,
			type: "image/png",
		});
		intent = Ti.Android.createIntentChooser(intent, "Choose");
		Ti.Android.currentActivity.startActivity(intent);
	});
});
window.add(button);
window.open();
```

---
**Test TabGroup with "Don't keep activities":**
1. Go to the Android device's main "Settings" screen.
2. Tap on "System" under "Settings.
3. Tap on "Developer options" under "System" settings.
4. Enable "Don't keep activities", which should be near the bottom of the list.
5. Build and run the below code on the above Android device.
6. Tap on the "Open Child Window" button.
7. Tap on the "Close Window" button.
8. Verify that you are returned back to a 3 tab window. _(Should not be blank, which was the bug.)_

```javascript
function createTab(title) {
	var window = Ti.UI.createWindow({ title: title });
	window.add(Ti.UI.createLabel({
		text: title + " View",
		top: "25%",
	}));
	var openButton = Ti.UI.createButton({
		title: "Open Child Window",
		top: "50%",
	});
	openButton.addEventListener("click", function() {
		var childWindow = Ti.UI.createWindow({
			title: title + " Child Window",
			backgroundColor: "white",
		});
		var closeButton = Ti.UI.createButton({
			title: "Close Window",
		});
		closeButton.addEventListener("click", function() {
			childWindow.close();
		});
		childWindow.add(closeButton);
		childWindow.open();
	});
	window.add(openButton);
	var tab = Ti.UI.createTab({
		title: title,
		window: window,
	});
	return tab;
}

var tabGroup = Ti.UI.createTabGroup({
	title: "TabGroup Parent Test",
	tabs: [createTab("Tab 1"), createTab("Tab 2"), createTab("Tab 3")],
});
tabGroup.open();
```

---
**Test LiveView Restarts with "Don't keep activities":**
1. Go to the Android device's main "Settings" screen.
2. Tap on "System" under "Settings.
3. Tap on "Developer options" under "System" settings.
4. Enable "Don't keep activities", which should be near the bottom of the list.
5. Build and run the below code on the above Android device.
6. Edit the "app.js" file and save your changes.
7. Verify that the app restarts and that the below `Ti.API.info()` message is logged again.

```javascript
Ti.API.info("### The 'app.js' file was loaded.");
var window = Ti.UI.createWindow();
window.open();
```

jquick-axway committed:
[TIMOB-26966] Android: Refactored activity/fragment restore handling

- [TIMOB-26966] Fixed bug where back navigating from child window causes app to exit when "Don't keep activities" is enabled as of 7.2.1.
- [TIMOB-26914] Fixed 8.0.0 regression where an OS forced quit app displays a blank window upon relaunch. (Was incorrectly restoring child activity.)
- [TIMOB-17089] Resolved view ID conflicts causes app crashes when activity/fragment is restored.
- [TIMOB-26964] Fixed bug where TabGroup would be blank (no tabs) when calling child window's close() method while "Don't keep activities" is enabled.
- [TIMOB-26890] Fixed bug where tapping 3rd tab in TabGroup can cause a crash due to view ID conflict during fragment restore.
- Increased TabGroup's ViewPager offscreen page limit from 1 to 128 to avoid tab fragment from being destroyed/restored.
- Added several safety mechanisms to AlertDialog and ProgressIndicator dialog code to avoid crashes.

build commented:
<!--
  1 failure:  :disappointed_rel...
  1 warning:  :mag: Can't find ...
  1 messages
  
  DangerID: danger-id-default;
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  :disappointed_relieved: `npm test` failed. See below for details.
  </td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  :mag: Can't find junit reports at `./junit.*.xml`, skipping generating JUnit Report.
  </td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  ```
> titanium-mobile@8.1.0 test /Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-10829
> npm run ios-sanity-check && grunt


> titanium-mobile@8.1.0 ios-sanity-check /Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-10829
> ./build/scons check-ios-toplevel

Running "appcJs:src:lintOnly" (appcJs) task

Running "eslint:src" (eslint) task

/Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-10829/iphone/cli/commands/_build.js
  6410:79  warning  'out' is defined but never used  no-unused-vars

/Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-10829/iphone/cli/hooks/frameworks.js
   61:5   warning  Expected catch() or return                  promise/catch-or-return
   61:34  warning  Avoid calling back inside of a promise      promise/no-callback-in-promise
  428:5   warning  Each then() should return a value or throw  promise/always-return

✖ 4 problems (0 errors, 4 warnings)


Running "checkFormat:ios" (checkFormat) task

Running "checkFormat:android" (checkFormat) task
Fatal error: Error: Formatting incorrect on "android/titanium/src/java/org/appcelerator/titanium/TiBaseActivity.java", proposed changes: <?xml version='1.0'?>
<replacements xml:space='preserve' incomplete_format='false'>
<replacement offset='49425' length='1'>&#10;					</replacement>
<replacement offset='49486' length='17'> </replacement>
</replacements>

npm ERR! Test failed.  See above for more details.

```
  </td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against ce69257685f23337c597ebb349ee1a9f05ec65f9
</p>

jquick-axway committed:
Android: Updated code formatting for [TIMOB-26966]

jquick-axway committed:
Merge branch 'master' into TIMOB-26966

jquick-axway committed:
Android: Fixed [TIMOB-26966] to not crash when putting a fragment within a fragment such as "ti.map" view

- Re-added ID assigned to TiUIFragment's view, but it's now guaranteed to be assigned a unique ID.

garymathews committed:
Merge branch 'master' into TIMOB-26966

jquick-axway committed:
[TIMOB-26975] Android: Fixed bug where LiveView does not reload app if "Don't keep activities" is enabled as of 8.0.0

jquick-axway committed:
Android: Added additional error checking to dialog code for [TIMOB-26966]

jquick-axway committed:
Android: Added additional dialog error checking to [TIMOB-26966]

- [TIMOB-15829] Fixed crash caused by progress indicator set to status bar that happens after hiding twice and then opening.

jquick-axway committed:
Merge branch 'master' into TIMOB-26966

ypbnv commented:
@jquick-axway TabGroup changes look good to me!

garymathews committed:
Merge branch 'master' into TIMOB-26966

sgtcoolguy committed:
Merge branch 'master' into TIMOB-26966

build commented:
<!--
  0 failure: 
  0 warning: 
  2 messages
  
  DangerID: danger-id-Danger;
-->



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  :floppy_disk: <a href="https://jenkins.appcelerator.org/job/titanium-sdk/job/titanium_mobile/job/PR-10829/13/artifact/dist/mobilesdk-8.1.0.v20190415083736-osx.zip">Here's the generated SDK zipfile</a>.
  </td>
    </tr>
  
<tr>
      <td>:book:</td>
      <td>

  :white_check_mark: All tests are passing
Nice one! All 3787 tests are passing.
  </td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against d1f5b71466bdf2a14c1cb82b38dcc7278af2cea3
</p>

garymathews committed:
Merge branch 'master' into TIMOB-26966

garymathews committed:
Merge branch 'master' into TIMOB-26966

garymathews commented on android/titanium/src/java/org/appcelerator/titanium/TiBaseActivity.java:
```diff
@@ -1726,6 +1684,22 @@ protected boolean shouldFinishRootActivity()
 	}
 
+	@Override
+	public void finishAfterTransition()
```
Is this override necessary, does't `finishAfterTransition` call `finish`?

jquick-axway replied:
The override is needed to call the `removeFromActivityStack(this)` like we do in the `finish()` method... because I removed the method call from the `WindowProxy` class. Now all of the `addToActivityStack()` and `removeFromActivityStack()` calls are done in the same place, the `TiBaseActivity` class. The old code wants to remove the activity from the stack (before the `onDestroy()` happens) so that the next child `Ti.UI.Window` that the code might open won't use its context.

garymathews reviewed (approved):
LGTM, just a minor comment above.

Since `onBackPressed` has been refactored, QE should also run tests from these tickets again just to be safe:
https://jira.appcelerator.org/browse/TIMOB-24531
https://jira.appcelerator.org/browse/TIMOB-23969
https://jira.appcelerator.org/browse/TIMOB-23989

lokeshchdhry committed:
Merge branch 'master' into TIMOB-26966

lokeshchdhry commented:
@jquick-axway , For https://jira.appcelerator.org/browse/TIMOB-24531. 
Test code here : https://github.com/appcelerator/titanium_mobile/pull/8906#issue-112469922
When I open the second window with `exitOnClose` set to `true` & press back I see the first window for like a second.
With 8.0.0 I don't see the first window at all. It directly quits.

garymathews commented:
@jquick-axway To address the issue @lokeshchdhry found, should we change the direction we terminate the activity stack?

https://github.com/appcelerator/titanium_mobile/pull/10829/files#diff-0d6c63e01d569acf469b905d84a1dacbR188

Terminate the child activities before the current?

jquick-axway committed:
Android: More improvements made to [TIMOB-26966]

- [TIMOB-26966] Fixed bug where Window/TabGroup close()method call is ignored until "open" event has been fired.
  * Resolved by modifying proxy's handleClose() to call TiActivityWindows.remove() immediately.
- Fixed bug where dynamically changing "exitOnClose" would be ignored.
  * Flag was only being read by activity's "launchIntent" in 8.0.0.
  * Now reads directly from proxy, which is the best solution. Especially if onNewIntent() was called. (Old code would have had a problem too.)
- Re-added support for "exitOnClose" in child windows. (Was only supported by root window.)
- Removed window exit animation if its "exitOnClose" was set true. (Prevents parent window briefly appearing upon exit.)
- Added more dialog safe-guards.
  * Would crash if failed to create dialog builder if there are no activities available.
  * Can happen if "Don't keep activities" is enabled, you press the Home button, and code attempts to show dialog in the background.

jquick-axway commented:
Updated PR.

ssjsamir commented:
@jquick-axway Please could you look at the merge conflict?

jquick-axway committed:
Merge branch 'master' into TIMOB-26966

jquick-axway committed:
Android: Updated code formatting for [TIMOB-26966]

jquick-axway committed:
Android: Modified [TIMOB-26966] to not exit app upon relaunch/restore after OS force quit it

garymathews reviewed (approved):
CR: **PASS**

lokeshchdhry commented:
FR Passed.

The activity/fragment refactor looks good. 

Studio Ver:     5.1.2.201903111843
SDK Ver:        8.1.0 local build
OS Ver:         10.14
Xcode Ver:      Xcode 10.1
Appc NPM:       4.2.13
Appc CLI:       7.0.10
Daemon Ver:     1.1.3
Ti CLI Ver:     5.1.1
Alloy Ver:      1.13.9
Node Ver:       8.15.1
NPM Ver:        6.4.1
Java Ver:       10.0.2
Devices:        ⇨ google Nexus 5 (Android 6.0.1)
                ⇨ google Pixel (Android 9)

lokeshchdhry committed:
Merge branch 'master' into TIMOB-26966
